vyper/context/types/utils.py

```diff
@@ -39,11 +35,18 @@ def get_type_from_abi(
     type_string = abi_type["type"]
     if type_string == "fixed168x10":
         type_string = "decimal"
+    # TODO string and bytes
 
     namespace = get_namespace()
     try:
+        if "[" in type_string:
+            value_type_string, length_str = type_string.rsplit("[", maxsplit=1)
+            length = int(length_str.rstrip("]"))
+            value_type = get_type_from_abi({"type": value_type_string}, is_constant=is_constant)
+            return ArrayDefinition(value_type, length, is_constant, is_public)
+
         return namespace[type_string]._type(is_constant=is_constant, is_public=is_public)
-    except KeyError:
+    except (KeyError, ValueError, InvalidType, UnknownType):
         raise UnknownType(f"ABI contains unknown type: {type_string}") from None
```

Comment:
Would be better if this was split it up better:
```python
    if "[" in type_string:
        value_type_string, length_str = type_string.rsplit("[", maxsplit=1)  # Shouldn't fail
        length_str = length_str.rstrip("]")
        if len(length_str) != 1:
            raise ...  # Invalid syntax
        try:
            length = int(length_str)  # Maybe also check if length has decimal places or something?
        except KeyError as e:
            raise ... from e  # Length is not an integer type
        if 0 > length or length < SOME_NUMBER:
           raise ...  # Length is not a valid range
        try:
            value_type = get_type_from_abi({"type": value_type_string}, is_constant=is_constant)
        except (KeyError, ValueError, InvalidType, UnknownType) as e:  # Not sure of the errors this throws
            raise UnknownType(f"ABI contains unknown type: {type_string}") from e
        return ArrayDefinition(value_type, length, is_constant, is_public)
    else:
        return namespace[type_string]._type(is_constant=is_constant, is_public=is_public)
```

All of this logic might live better in it's own function though

Reply:
* `if len(length_str) != 1:` - this would block arrays of a length of 10 or more
* the bounds checks on length are in `ArrayDefinition`

Reply:
I made the edits here - https://github.com/vyperlang/vyper/pull/2017/commits/76df4088b9f4c69b0016a493c459bb86ffc72daa

I don't like it though, too much duplicated code.  The exceptions being raised are different depending on the line, the but reason is always near-identical. I think a catchall exception handler is a much cleaner approach.

Reply:
It gives more specific error messages, so why does it hurt?

Reply:
Doesn't hurt.  Lets go with it.